src/compiler/checker.ts

```diff
@@ -34271,7 +34271,7 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         if (pos < paramCount) {
             return getTypeOfParameter(signature.parameters[pos]);
         }
-        if (signatureHasRestParameter(signature)) {
+        if (signature.parameters.length !== 0 && signatureHasRestParameter(signature)) {

```

Comment:
In a few lines above this, where `paramCount` is assigned, `signature.parameter` is `[]` and  `signatureHasRestParameter` is returned as true so `paramCount` is -1. Then it enters the if statement (because signature exists), but `signature.parameter.length` is 0, so it crashes when `getTypeOfSymbol(signature.parameters[paramCount])` is called.

Reply:
I see. What I'm wondering is whether it's correct that `signatureHasRestParameter(signature)` returns true (i.e. why `signature.flags` has `SignatureFlags.HasRestParameter` set) when `signature.parameter` is empty.

Reply:
It does sound to me like a signature with HasRestParameter must have at least one parameter, and that the problem is with the creation of the signature.

There are 4 cases that I can see which would cause HasRestParameter to get set:

1. The signature actually has a rest parameter.
2. In JS, for Closure Compiler syntax for signatures: `function(...)` (I can't remember the actual syntax).
3. In JS, when `arguments` is used in the body of a function.
4. In createUnionOfSignaturesForOverloadFailure, when any candidate signature has a rest parameter.

(2) and (3) are out because there's no JS here. I don't see any overloads for (4). For (1), `(...args: Args) => R` does have a rest parameter, and should have a parameter list. Can you check the crashy signature in the debugger to see if it's that one, or something based on that one. Maybe the conditional type or mapped type somehow incorrectly fails to map the parameter list. You might check all the assignments to signature.flags and make sure they're paired with assignments to signature.parameters.